### PR TITLE
Initial blog post

### PR DESCRIPTION
**Pull request by ambaxter**

### CODE REVIEW

File: blog.md [hunk 1]
```diff
@@ -0,0 +1,258 @@
+```
+ % ./target/release/bench -w seq -r seq
+# Write 64.945474ms     (64.945µs/op)   (15397 op/sec)
+# Read  1.000046609s    (24ns/op)       (41666666 op/sec)
+```
+
+With those two lines the last few months of labor bear fruit.
```
Comment (gjbianco) [inline blog.md]:
```suggestion
With those results, the last few months of labor bear fruit. 
```

File: blog.md [hunk 2]
```diff
@@ -0,0 +1,258 @@
+```
+ % ./target/release/bench -w seq -r seq
+# Write 64.945474ms     (64.945µs/op)   (15397 op/sec)
+# Read  1.000046609s    (24ns/op)       (41666666 op/sec)
+```
+
+With those two lines the last few months of labor bear fruit. 
+I reimplemented the BBolt database in Rust.
```
Comment (gjbianco) [inline blog.md]:
```suggestion
Over those months, I have reimplemented the Go-based [BBolt key-value database](https://github.com/etcd-io/bbolt) in Rust. 
```

File: blog.md [hunk 3]
```diff
@@ -0,0 +1,258 @@
+```
+ % ./target/release/bench -w seq -r seq
+# Write 64.945474ms     (64.945µs/op)   (15397 op/sec)
+# Read  1.000046609s    (24ns/op)       (41666666 op/sec)
+```
+
+With those two lines the last few months of labor bear fruit. 
+I reimplemented the BBolt database in Rust. 
+It is the culmination of my journey where I learned how to build a database, how to design database architecture in Rust, and how to use API design to remove footguns.
```
Comment (gjbianco) [inline blog.md]:
```suggestion
This marks the culmination of my journey where I learned how to:
* Build a database
* Design a Rust-centric database architecturet
* How to use API design to remove ["footguns"](https://en.wiktionary.org/wiki/footgun)
```

File: blog.md [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+
+# Impetus
+
+The project started last year when I wanted to learn how ETCD and the Raft algorithm worked which then lead me to learning how the BBolt database worked.
```
Comment (gjbianco) [inline blog.md]:
```suggestion
The project started in late 2023 when I wanted to learn how ETCD and the Raft algorithm work.
This also led me to learn how the BBolt database works. 
```

File: blog.md [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+```
+
+With those two lines the last few months of labor bear fruit. 
+I reimplemented the BBolt database in Rust. 
+It is the culmination of my journey where I learned how to build a database, how to design database architecture in Rust, and how to use API design to remove footguns.
+
+# Impetus
+
+The project started last year when I wanted to learn how ETCD and the Raft algorithm worked which then lead me to learning how the BBolt database worked. 
+Following in the software engineering tradition what started as curiosity turned into an all-consuming case of yak shaving.
```
Comment (gjbianco) [inline blog.md]:
```suggestion
I proceeded to follow in the software-engineering tradition of allowing curiosity to turn into an all-consuming case of [yak shaving](https://en.wiktionary.org/wiki/yak_shaving). 
```

File: blog.md [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+
+With those two lines the last few months of labor bear fruit. 
+I reimplemented the BBolt database in Rust. 
+It is the culmination of my journey where I learned how to build a database, how to design database architecture in Rust, and how to use API design to remove footguns.
+
+# Impetus
+
+The project started last year when I wanted to learn how ETCD and the Raft algorithm worked which then lead me to learning how the BBolt database worked. 
+Following in the software engineering tradition what started as curiosity turned into an all-consuming case of yak shaving. 
+
```
Comment (gjbianco) [inline blog.md:19]: I would expand a little more on the impetus here beyond just "idle curiosity"/"I was bored" lol. Specifically, which were the technical challenges that piqued your curiosity that produced the yak shaving?

File: blog.md [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+# Goals and Standards
```
Comment (gjbianco) [inline blog.md]:
```suggestion
# Project Standards
```

File: blog.md [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+* Built in stable Rust
+* Library API as close to the original as possible
+* On-file representation exactly matches original code 
+* Manage transaction memory without Rc
+* Unsafe behavior validated by Miri
```
Comment (gjbianco) [inline blog.md]:
```suggestion
* Use only stable Rust
* Keep the library API as close to the original as possible
* The on-file representation _must_ match original code exactly
* Manage transaction memory without Rc
* Unsafe behavior validated by Miri
```
Links to Rc and Miri?

File: blog.md [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+
+# BBolt Under a Microscope
```
Comment (gjbianco) [inline blog.md:215]: I really like this heading.

File: blog.md [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+Under the covers BBolt is a minimalistic B+ tree database designed to store Buckets.
```
Comment (gjbianco) [inline blog.md]:
```suggestion
Under the covers, BBolt is a minimalistic B+ tree database designed to store buckets. 
```
If BBolt has a named object called "Bucket", I would instead maybe stylize it as `Bucket` (surround with backticks).

File: blog.md [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+# Goals and Standards
+* Built in stable Rust
+* Library API as close to the original as possible
+* On-file representation exactly matches original code 
+* Manage transaction memory without Rc
+* Unsafe behavior validated by Miri
+
+# BBolt Under a Microscope
+Under the covers BBolt is a minimalistic B+ tree database designed to store Buckets. 
+Buckets may store key/value pairs in bytes and sub-Buckets.
```
Comment (gjbianco) [inline blog.md]:
```suggestion
_Buckets_ store key-value pairs in bytes and sub-Buckets. 
```
Again, consider backticks.

File: blog.md [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+* Built in stable Rust
+* Library API as close to the original as possible
+* On-file representation exactly matches original code 
+* Manage transaction memory without Rc
+* Unsafe behavior validated by Miri
+
+# BBolt Under a Microscope
+Under the covers BBolt is a minimalistic B+ tree database designed to store Buckets. 
+Buckets may store key/value pairs in bytes and sub-Buckets. 
+Sub-buckets may be stored inline inside a parent bucket or on its own page.
```
Comment (gjbianco) [inline blog.md]:
```suggestion
Sub-buckets live inline within a parent bucket or on their own page.
```
It might be better to flip this around and make "parent bucket" the subject in order to more-easily avoid passive tense.

It's unclear (to me) what a page means here.

File: blog.md [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+* On-file representation exactly matches original code 
+* Manage transaction memory without Rc
+* Unsafe behavior validated by Miri
+
+# BBolt Under a Microscope
+Under the covers BBolt is a minimalistic B+ tree database designed to store Buckets. 
+Buckets may store key/value pairs in bytes and sub-Buckets. 
+Sub-buckets may be stored inline inside a parent bucket or on its own page.  
+
+The data types used to represent the BBolt database are relatively straightforward.
```
Comment (gjbianco) [inline blog.md]:
```suggestion
The data types that represent the BBolt database are relatively straightforward.
For example, the following <thing> <verbs> [a <another thing>].
```

I'm not sure saying it is straightforward tells the reader anything. It's either a self-evident fact or it runs the risk that, if they _don't_ find it straightforward, that the rest of the article is going to be over their head and they should quit reading.

File: blog.md [hunk 14]
```diff
@@ -0,0 +1,258 @@
...
+* Unsafe behavior validated by Miri
+
+# BBolt Under a Microscope
+Under the covers BBolt is a minimalistic B+ tree database designed to store Buckets. 
+Buckets may store key/value pairs in bytes and sub-Buckets. 
+Sub-buckets may be stored inline inside a parent bucket or on its own page.  
+
+The data types used to represent the BBolt database are relatively straightforward. 
+
+```rust
```
Comment (gjbianco) [inline blog.md:222]: I would break this up into multiple code snippets with fewer inline comments and more plain-English explanations of what's going on. Don't feel like you need to explain all of Rust, but don't assume that your audience are all Rust experts.

Also, if there are key parts of this code (perhaps the bigger structs?), then maybe only include and talk about those. You can handwave away a few things here if it improves overall clarity.

File: blog.md [hunk 15]
```diff
@@ -0,0 +1,258 @@
...
+/// The Page ID. Page address = `PgId` * meta.page_size
+#[repr(C)]
+pub struct PgId(u64);
+
+/// The Transaction ID. Monotonic and incremented every commit
+#[repr(C)]
+pub struct TxId(u64);
+
+/// `PageHeader` represents the on-file layout of a page header.
```
Comment (gjbianco) [inline blog.md:231]: This is a prime example. I would _hope_ that something called a "page header" represents a "page header" lol. What is a page header? Why do I care about it?

File: blog.md [hunk 16]
```diff
@@ -0,0 +1,258 @@
...
+  id: PgId,
+  /// Page's type. Branch(0x01), Leaf(0x02), Meta(0x04), or FreeList(0x10)
+  flags: u16,
+  /// Defines the number of items in the Branch, Leaf, and Freelist pages
+  count: u16,
+  /// How many additional meta.page_size pages are included in this page
+  overflow: u32,
+}
+
+/// `BucketHeader` represents the on-file layout of a bucket header.
```
Comment (gjbianco) [inline blog.md:246]: This is a bit better. I would pull this out of a codeblock and do a little more explanation in plain English.

File: blog.md [hunk 17]
```diff
@@ -0,0 +1,258 @@
...
+  pos: u32,
+  /// Key length
+  key_size: u32,
+  /// Page ID of this branch's
+  pgid: PgId,
+}
+
+```
+
+The 4 database page types are also straightforward.
```
Comment (gjbianco) [inline blog.md:314]: Similar as before. Don't necessarily assume _all_ of your audience is intimately familiar with database architecture.

Here, look to provide explanations that would answer questions like: what is a page? Why are there different types? Why do I care about those types.

File: blog.md [hunk 18]
```diff
@@ -0,0 +1,258 @@
...
+  pgid: PgId,
+}
+
+```
+
+The 4 database page types are also straightforward. 
+
+Page Layouts
+```
+      Meta Page
```
Comment (gjbianco) [inline blog.md:318]: Similar to the codeblock before, break each of these (really nice) diagrams into separate codeblocks and provide a blurb for each, with transitions/captions.

The goal is to be able to skim the article and get an idea of what's contained therein. Naturally, if I see largely two big codeblocks with very little English, I'm inclined to just assume it's a dump of some code and lose interest.

File: blog.md [hunk 19]
```diff
@@ -0,0 +1,258 @@
...
+      │Key 0 │◀──── elem ptr + elem.pos
+      ├──────┤
+      │Key 1 │
+      ├──────┤
+      │Key 2 │
+      ├──────┤
+      │ .... │
+      └──────┘
+
+Created with Monodraw
```
Comment (gjbianco) [inline blog.md]: Good attribution, but pull it out of the codeblocks and/or provide the info in footer. It's essentially tertiary information.

File: blog.md [hunk 20]
```diff
@@ -0,0 +1,258 @@
...
+```
+
+# Foundations of Reading Mmapped Files
```
Comment (gjbianco) [inline blog.md:387]: Good heading, though I do wonder if this section should be higher. I also immediately wonder "what is an mmapped file and why do I care?" You basically have to constantly provide reasons for your reader to keep reading.

File: blog.md [hunk 21]
```diff
@@ -0,0 +1,258 @@
...
+[memmap2](https://crates.io/crates/memmap2)
+[Mmap](https://en.wikipedia.org/wiki/Mmap) [CreateFileMappingW](https://learn.microsoft.com/en-us/windows/win32/api/memoryapi/nf-memoryapi-createfilemappingw)
```
Comment (gjbianco) [inline blog.md:389]: Good attributions! I would have a `# References` section at the bottom with a lot of these types of links though (including Monodraw).

File: blog.md [hunk 22]
```diff
@@ -0,0 +1,258 @@
...
+
+## Reading a Page
```
Comment (gjbianco) [inline blog.md:391]: Is "reading" the best verb here? Who is doing the reading?

File: blog.md [hunk 23]
```diff
@@ -0,0 +1,258 @@
...
+
+## Reading a Meta Page
+
+```go
```
Comment (gjbianco) [inline blog.md:395]: You probably know what I'm going to say here :) English explanations!

File: blog.md [hunk 24]
```diff
@@ -0,0 +1,258 @@
...
+// key returns a byte slice of the node key.
+func (n *leafPageElement) key() []byte {
+	i := int(n.pos)
+	j := i + int(n.ksize)
+	return unsafeByteSlice(unsafe.Pointer(n), 0, i, j)
+}
+```
+
+# Removing Footguns With API Design
```
Comment (gjbianco) [inline blog.md]:
```suggestion
# Hardening the API at the Design Level
```

I would reserve using cutesy jargon terms like "footgun" in the paragraphs themselves, not headers. Again, assume most readers are going to skim before reading.

**Review by gjbianco — COMMENTED** [review]
In addition to the in-line comments and suggestions, I urge you to organize what you've got (and add) into, very loosely:

1. High-level explanations about what it does and why (you do this a bit with the impetus section, but I also mean the why's for technial decisions, not just "why I did this"). As is, it's a bit unclear when to expect a technical explanation vs a strategic one.
2. Technical details. Basically, you want the article to flow from highest-level to lowest-level. Some readers inherently won't be interested in the technicalities, so you want to provide a clearer cutoff point for when they quit reading. Otherwise, they will stop reading prematurely (i.e. the moment they see too much jargon or codeblocks). That doesn't detract from your largely technical audience, either -- it likely just helps even those folks consume the article easier.

File: blog.md (hunk 20 above)
Comment (ambaxter) [reply blog.md:387]: The mmapped file and why do I care is definitely coming. I got pulled into a meeting before I could get too much further here

File: blog.md (hunk 21 above)
Comment (ambaxter) [reply blog.md:389]: I will be using these links in my description of mmap

File: blog.md (hunk 16 above)
Comment (ambaxter) [reply blog.md:246]: These are direct copies of the struct definitions. Would it not be better to leave the descriptions as they are in the code?
